### ISSUE TITLE
VNNCOMP2022 Report and Discussion for Next Year

### ISSUE THREAD

**Issue by stanleybak** [stanleybak/vnncomp2022#4]
The scoring repo with results is available [here](https://github.com/ChristopherBrix/vnncomp2022_results). Please check the `SCORING/results.txt` file [here](https://github.com/ChristopherBrix/vnncomp2022_results/blob/master/SCORING/results.txt) (generated by [process_results.py](https://github.com/ChristopherBrix/vnncomp2022_results/blob/master/SCORING/process_results.py)) for any errors and let us know if you find anything.

We will send information about the report in the coming days, so you can write a few paragraphs about your tool / benchmark. We'll post the overleaf location in the mailing list.

The presentation recording is available [here](https://drive.google.com/file/d/1JQ_pmTeJCuRhMT_ubgbQnwgruCJaohad/view?usp=sharing) and the slides [here](https://drive.google.com/file/d/1nnRWSq3plsPvOT3V-drAF5D8zWGu02VF/view?usp=sharing). We're working on getting a higher quality recording and plan to post it in somewhere public like CAV's youtube channel.

### Ideas for Next Year
Please use this issue to continue the discussion about changes that we want for the next edition. The notes from the slides are:

* Disallow tool tuning per benchmark and instead focus on automated strategy selection: `if (#inputs > 100) set_settings_image(); else set_settings_control();`  This may be more Beneficial for real-world users
*Require all benchmarks to have at least one sat and one unsat instance
*Filter out / reduce scores for easy sat instances (e.g. via adv attack)
*Handling on fast/easy instances:
* Minimum execution time < 1 second?
* Don’t subtract tool overhead?
* Batch mode verification?
* Separate category for falsification
* Check all counterexamples for correctness for scoring, not just when
tools don’t agree
* Look into effect of (floating point) precision for analysis
* Look into other Hardware/AWS instance options

**Comment by stanleybak** [comment]
One improvement discussed would be to create a better general VNN-LIB parser library (in python?) that people could use. I think the code many teams use based on `simple_adversarial_generator` was not meant to be a long-term implementation. I had some issues with VGGNET for this, where any-category misclassification with 1000 output categories took forever to parse. There were also certain specs I couldn't express, like a PDG attack vector multiplied by some percentage, or a brightening attack where all pixels get brightened at the same rate.

**Comment by piyush-J** [comment]
Regarding [one of our questions](https://github.com/stanleybak/vnncomp2022/issues/1#issuecomment-1113805809) we asked a few months back:

> Our group has been building an algorithm selection-driven meta-solver, i.e., a solver that leverages machine learning to determine which algorithms to use for a particular instance. These techniques are extremely powerful in practice. However, they are often controversial in competition environments.
> 
> However, "static" algorithm selection seems to exist in several solvers from last year on benchmark categories. Specifically, several solvers I noticed had different configurations depending on the type of benchmark, since it was known beforehand. 

We received a [reply](https://github.com/stanleybak/vnncomp2022/issues/1#issuecomment-1116266349) saying that we can discuss this for the next iteration of the competition. 

We request you to consider the possibility of including meta-solvers from next year. 

Thank you!

**Comment by huanzhang12** [comment]
Here are my two cents:

1. The automated website this year was great! I hope we can have a setup script for benchmarking generation as well, similar to the tool setup script, to make sure benchmarks can always be correctly produced;
2. Define a binary format for saving counterexamples, e.g., a flattened numpy array, to allow efficient saving of counterexamples with large dimensions.

Thanks you for organizing this great competition and making it better each year!

**Comment by stanleybak** [comment]
I wonder how people think we should handle batch-execution benchmarks, if we wanted to add this. 

Should people still propose a .csv file with individual timeouts, or we just have a global timeout and let the tool figure which instances to allocate computation time. The latter is probably more realistic right? But then tools could get stuck on hard instances and exhaust their global timeout.

It might be preferable to have this be reverse compatible, so that tools without batch-mode supported could still participate, they just have overhead counted against their verification time. So maybe the tool produces one result file per instance? 

I also wonder if we should facilitate restarting if a tool crashed at any point, or ran out of memory. This could require more complex management scripts though.

**Comment by ChristopherBrix** [comment]
> * The automated website this year was great! I hope we can have a setup script for benchmarking generation as well, similar to the tool setup script, to make sure benchmarks can always be correctly produced;

I'm happy to hear that, thank you! I will definitely look into including a setup script next time.



> 2\. Define a binary format for saving counterexamples, e.g., a flattened numpy array, to allow efficient saving of counterexamples with large dimensions.

That sounds reasonable, zipped plain text files aren't ideal.

**Comment by stanleybak** [comment]
Here are some further notes on potential changes for next time:

* How can we encourage general strategies and not extensive per-network tuning?
* We can filter difficulty at generation time or provide feedback when submitting
* Configuration files (for things like 'run install as sudo and other options') could be moved from web form to github yaml (or other) files, so that they're public and version controlled
* exclude benchmarks only in prepare_instance.sh (we missed some benchmarks that should have been skipped initially)
* Should we continue to allow closed source tools? There's no way to check if they abuse overhead / timeouts.
* How should we handle meta solvers? Explicit category?
* How to handle potentially unsound tools? Require all benchmarks should have some SAT instances?
* split benchmarks into "standard" ones and others. Authors spent lots of time supporting non-standard benchmarks and it's not effort that generalizes the tool at all but very specific for the benchmarks we had (for example segmentation benchmarks with argmax layers)

**Comment by ttj** [comment]
Here is a read-only link to the 2022 report, to be updated, and all participants should add description of their tools and/or benchmarks. I have sent invites to an editable version to most emails I had for participants. If anyone should be added to be able to edit, please let me know.

https://www.overleaf.com/read/pfcwgdvyyrwk

**Comment by pomodoromjy** [comment]
> Here is a read-only link to the 2022 report, to be updated, and all participants should add description of their tools and/or benchmarks. I have sent invites to an editable version to most emails I had for participants. If anyone should be added to be able to edit, please let me know.
> 
> https://www.overleaf.com/read/pfcwgdvyyrwk

We will edit benchmark "Carvana_unet". The email address is luoyg3@changan.com.cn Could you add us into the list?

**Comment by huanzhang12** [comment]
@stanleybak These are very good suggestions for potential future changes, and here are my thoughts on some of your notes:

> * How can we encourage general strategies and not extensive per-network tuning?

One way I thought about is to propose a benchmark with randomly generate network architecutures during testing (e.g., the number of layers, neurons, etc are random with different seeds), so the tuning must be general for a large range of networks.

> * split benchmarks into "standard" ones and others. Authors spent lots of time supporting non-standard benchmarks and it's not effort that generalizes the tool at all but very specific for the benchmarks we had (for example segmentation benchmarks with argmax layers)

I also agree that the image segmentation benchmark is quite different from others and can take a lot of time, but I feel it is probably not easy to make the split right. On one hand, it is subjective to tell which benchmark is "standard" and which one is not, given that all of them are using standard onnx and vnnlib format. On the other hand, if we do not score "non-standard" benchmarks I guess teams will have no incentive to work on them due to their hardness, so essentially they will be excluded from the competition.

> * Should we continue to allow closed source tools? There's no way to check if they abuse overhead / timeouts.

I am also concerned about closed source tools. I guess at least the organizers should have a way to verify they are not abusing anything. Ideally the code should be made available to organizers under some NDA.

> * How should we handle meta solvers? Explicit category?

An explicit category is a good idea, and I think a high-performance and user-friendly meta solver can be very helpful for making verification more applicable to real scenarios, so we should encourage meta-solver builders to participate.

> * How to handle potentially unsound tools? Require all benchmarks should have some SAT instances?

I like the idea of requiring each benchmark to have at least a few percent of SAT instances, which can be useful to detect unsound tools. We should also make sure these SAT instances are randomized enough.

**Comment by hhijazi** [comment]
Is there a timeline for VNN 2023? We're interested in participating and bringing instances from another competition: https://gocompetition.energy.gov/ 
This will hopefully improve the diversity of the benchmarks that seem to currently focus on image classification.

**Comment by stanleybak** [comment]
@hhijazi Yes more variety of benchmarks would be good. Do those competitions have input / output specifications as well? Usually we start discussing the competition around Feb/March, so your timing is pretty good.

**Comment by hhijazi** [comment]
The competition itself is not for NN verification but the data can be used to evaluate NNs that try to find the optimal solution to a power system problem. They do have specifications for input/output for the different challenges, e.g.: https://gocompetition.energy.gov/challenges/challenge-2/input-data-format

**Comment by stanleybak** [comment]
Sounds relevant. For VNN-COMP, the benchmark proposer would need to create specific NNs to evaluate as well as provide input / output specifications. Last year we required this had some randomness to it, but as a first step just preparing one specific network and spec is sufficient.

**Comment by hhijazi** [comment]
Sounds good, where do we submit these and what is a hard deadline for doing so?

**Comment by stanleybak** [comment]
Last year this was organized around[ this github issue](https://github.com/stanleybak/vnncomp2022/issues/2), and we had a website for submissions. This was done around April/May and I expect a similar timeline this year.

**Comment by hhijazi** [comment]
Got it, thanks Stanley.

**Comment by hhijazi** [comment]
Quick question: did the organizers consider putting the competition on Kaggle to increase its visibility?

**Comment by stanleybak** [comment]
I don't think we did as verification is not really a machine learning technique, but maybe the scope of Kaggle is something we've underestimated.

**Comment by hhijazi** [comment]
I believe it would really increase participation and impact, Kaggle has hosted non-ML competitions in the past, for instance, they have a yearly Optimization competition around Christmas time: https://www.kaggle.com/competitions/santa-2022
(this one had 874 teams!)

**Comment by stanleybak** [comment]
@hhijazi (and anyone else who may be interested) We would like to invite you to participate in the 4th Verification of Neural Networks Competition (VNN-COMP) (https://sites.google.com/view/vnn2023). 

Registration: We ask that you indicate your interest by March 17, 2023, if you are interested to learn more, and potentially participate, in order to help guide the direction of the competition. The registration form is here: https://forms.gle/tjPzsSP4qNfzAMtZ7
